main/src/library/Channel/Channel.flix

```diff
@@ -0,0 +1,582 @@
+    let convertedStr = toString(val);
+    StringBuilder.appendString!(stringBuilder, str);
+    StringBuilder.appendString!(stringBuilder, convertedStr);
+    let fullStr = StringBuilder.toString(stringBuilder);
+    Console.printLine(fullStr)
+
+
+namespace Channel {
+
+    // todo: remove this
+    def exit(status: Int32): () & Impure =
+        import java.lang.Runtime:getRuntime();
+        let runtime = getRuntime();
+        import java.lang.Runtime.exit(Int32);
+        runtime.exit(status)
+
+    /**
+    * The tuple corresponds to (id, channelLock, bufferSize, unbuffered, elementQueue, waitingGetters, waitingSetters)
+    *   - id
+    *   - channelLock is the channelLock of this channel.
+    *   - bufferSize is the size of a channel. If you try to put an
+    *   element in a channel that's full, you wait until there's space.
+    *   - a flag for whether the channel is unbuffered or not
+    *   - elementQueue is the elementQueue of elements in the list.
+    *   - waitingGetters is a set of conditions that is waiting for get.
+    *   This set is cleared after each new element.
+    *   - waitingSetters is a condition that can notify threads of
+    *   available space in the elementQueue.
+    */
+    pub opaque type Channel[a] = (
+        Int64,
+        ReentrantLock,
+        Int,
+        Bool,
+        MutList[a],
+        MutList[LockConditionPair],
+        Condition
+    )
+
+    /**
+    * public int branchNumber;
+    * public boolean defaultChoice;
+    * public Object element;
+    */
+    pub opaque type SelectChoice[a] = (
+        Int,
+        Bool,
+        a
+    )
+
+    pub def newSelectChoice(branchNumber: Int, element: a, defaultChoice: Bool): SelectChoice[a] =
+        SelectChoice(
+            branchNumber,
+            defaultChoice,
+            element
+        )
+    // FYI default choice is false in this constructor, TODO : delete
+
+    // channel, index
+    pub opaque type ChannelIndexPair[a] = (
+        Channel[a],
+        Int
+        )
+
+    pub def newChannelIndexPair(channel: Channel[a], index: Int): ChannelIndexPair[a] =
+        ChannelIndexPair(
+            channel,
+            index
+        )
+
+    pub def getChannel(cip: ChannelIndexPair[a]): Channel[a] = match cip {
+            case ChannelIndexPair(c, _)   => c
+        }
+
+    pub def getIndex(cip: ChannelIndexPair[a]): Int = match cip {
+            case ChannelIndexPair(_, ind)   => ind
+        }
+
+    /// @Override
+    /// pub def equals(o: a): Bool & Impure =
+    ///     if (this == o) return true;
+    ///     if (o == null || getClass() != o.getClass()) return false;
+    ///     ChannelIndexPair that = (ChannelIndexPair) o;
+    ///     return index == that.index && Objects.equals(channel, that.channel);
+    /// }
+
+    /// @Override
+    /// pub def hashCode(cip: ChannelIndexPair): Int =
+    ///     import java.util.Objects.hash as hash;
+    ///     hash(getChannel(cip), getIndex(cip))
+
```

Comment:
Indeed, you should define instances for: `ToString` and perhaps `Eq` and `Order` for your `ChannelImpl` type.